src/epoch/participants.rs

```diff
@@ -0,0 +1,146 @@
+// Global participant list.
+//
+// The global participant list is an intrustive list in which items are lazily removed on traversal
+// (after being "logically" deleted by becoming inactive).
+
+use std::{ops, mem};
+use std::sync::atomic;
+
+use epoch::{Atomic, Guard};
+use epoch::participant::Participant;
+use CachePadded;
+
+/// A participant node.
+///
+/// The data is cache padded to avoid cache line racing.
+#[derive(Debug)]
+pub struct ParticipantNode(CachePadded<Participant>);
+
+impl Default for ParticipantNode {
+    fn default() -> ParticipantNode {
+        ParticipantNode(CachePadded::new(Participant::default()))
+    }
+}
+
+impl ops::Deref for ParticipantNode {
+    type Target = Participant;
+    fn deref(&self) -> &Participant {
+        &self.0
+    }
+}
+
+impl ops::DerefMut for ParticipantNode {
+    fn deref_mut(&mut self) -> &mut Participant {
+        &mut self.0
+    }
+}
+
+/// A thread-safe list of threads participating in epoch management.
+///
+/// Currently implemented like a Treiber stack.
+#[derive(Debug, Default)]
+pub struct Participants {
+    /// The head node.
+    head: Atomic<ParticipantNode>,
+}
+
+impl Participants {
+    // TODO: Remove this cfg when `const fn` is stabilized.
+    /// Create an empty list of participants.
+    #[cfg(feature = "nightly")]
+    pub const fn new() -> Participants {
+        Participants { head: Atomic::null() }
+    }
+
+    /// Enroll the current thread into the list.
+    ///
+    /// This adds the current thread to the list in epoch management by adding a new `Particpant`
+    /// record to the global list.
+    pub fn enroll(&self) -> *const Participant {
+        // The new list.
+        let mut participant = Box::new(ParticipantNode::default());
+
+        // We ultimately use epoch tracking to free `Participant` nodes, but we can't actually
+        // enter an epoch here, so fake it; we know the node can't be removed until marked inactive
+        // anyway.
+        let fake_guard = ();
+        let guard: &'static Guard = unsafe { mem::transmute(&fake_guard) };
+
+        loop {
+            // Load the head of the participant list we're traversing.
+            let head = self.head.load(atomic::Ordering::Relaxed, guard);
```

Comment:
Like with ConcBag we load head twice in this loop. The compare and swap already returns the new value so we don't need to load head again. We should also use a pause instruction on x86.